### DISCUSSION TITLE
Error "XML Error: Element size must be a multiple of dim+1 mujoco"

### DISCUSSION THREAD

**Discussion by Nick20500** [google-deepmind/mujoco#2107] [Asking for Help] [answered]
### Intro

Hi!

I am a  undergrad student at ETH Zurich, I use MuJoCo for my research on Multimaterial deformable objects.

### My setup

I use MuJoCo 3.2.3 with python bindings.

### My question

When loading the file "parts_3800_3.xml" in the mujoco viewer, the error 
"XML Error: Element size must be a multiple of dim+1 mujoco
Element 'flexcomp', line 46" 
is displayed.

This error should imply, that the file "part-1.41.msh" has an error, i.e. one of the $Elements$ hasn't the right format/dimension.
But after manual inspection, manual comparison of the files "part-0.41.msh" and "part-1.41.msh" and automated inspection via python script, I can't find an error in the "part-1.41.msh" file, nore a difference in the format between the two files. And by the mujoco viewer, the file "part-0.41.msh" is correct since I can view it and the error only appears for the "part-1.41.msh" not for the "part-0.41.msh".

When I remove the flexcomp with the name "part 2", I can view the "parts_3800_3.xml" without any error in the mujoco viewer.
I am also able to load all 3 .msh files referenced in the "parts_3800_3.xml" file, in the gmsh viewer without any problems.

Is there actually a problem with the file "part-1.41.msh" which I and my code have overlooked or is there a bug in mujoco?

I have attached all 4 files in the follwoing zip folder:

[mujoco_files_from_issue.zip](https://github.com/user-attachments/files/17192119/mujoco_files_from_issue.zip)

Thank you for your help.

### Minimal model and/or code that explain my question

Model:
[mujoco_files_from_issue.zip](https://github.com/user-attachments/files/17192384/mujoco_files_from_issue.zip)

```XML

<!-- 3 PARTS
    
    created by Sebastian Heckers on 25.04.2024
    
    Example of multi-material deformable objects with 3 parts/meshes.
    With Bottle (ID 3800), body, neck and lid.
-->

<mujoco model="Sphere Drop">

    <include file="../drop/scene.xml"/>


    <asset>
        <texture name="tex" type="2d" builtin="checker" width="512" height="512" rgb1=".1 .1 .9" rgb2=".9 .3 .3"/>
        <material name="tex" texture="tex" texrepeat="1 1" texuniform="true" reflectance=".2"/>
    </asset>

    <!-- set default units -->
    <compiler angle="radian" meshdir="../../../assets/"/>

    <extension>
        <plugin plugin="mujoco.elasticity.solid" />
    </extension>

    <compiler autolimits="true"/>

    <option solver="CG" tolerance="1e-6" timestep=".0001"/>

    <size memory="240M"/>

    <worldbody>

        <flexcomp name="part1" type="gmsh" file="partnet/3800/part-0.41.msh" rgba=".8 .2 .2 1" pos="0 0 5" euler="0 0 -1.57075" mass="500" scale="3 3 3" >
            <!-- <contact condim="3" solref="0.01 1" solimp=".95 .99 .0001" selfcollide="none" friction="5 0.5 0.01"/> -->
        
            <plugin plugin="mujoco.elasticity.solid">
                <!--Units are in Pa (SI)-->
                <config key="young" value="1e5"/>
                <config key="poisson" value="0.45"/>
                <config key="damping" value="0.0"/>
            </plugin>

        </flexcomp>

        <flexcomp name="part2" type="gmsh" file="partnet/3800/part-1.41.msh" rgba=".2 .2 .8 1" pos="0 0 5" euler="0 0 -1.57075" mass="500" scale="3 3 3" >
            <!-- <contact condim="3" solref="0.01 1" solimp=".95 .99 .0001" selfcollide="none" friction="5 0.5 0.01"/> -->
        
            <plugin plugin="mujoco.elasticity.solid">
                <!--Units are in Pa (SI)-->
                <config key="young" value="1e7"/>
                <config key="poisson" value="0.45"/>
                <config key="damping" value="0.0"/>
            </plugin>

        </flexcomp>

        <flexcomp name="part3" type="gmsh" file="partnet/3800/part-2.41.msh" rgba=".2 .8 .2 1" pos="0 0 5" euler="0 0 -1.57075" mass="500" scale="3 3 3" >
            <!-- <contact condim="3" solref="0.01 1" solimp=".95 .99 .0001" selfcollide="none" friction="5 0.5 0.01"/> -->
        
            <plugin plugin="mujoco.elasticity.solid">
                <!--Units are in Pa (SI)-->
                <config key="young" value="1e8"/>
                <config key="poisson" value="0.45"/>
                <config key="damping" value="0.0"/>
            </plugin>

        </flexcomp>

    </worldbody>

</mujoco>



```
</details>

Code:
```python
import mujoco
import mediapy as media
model = mujoco.MjModel.from_xml_string(xml)
data = mujoco.MjData(model)

with mujoco.Renderer(model) as renderer:
  mujoco.mj_forward(model, data)
  renderer.update_scene(data)

  media.show_image(renderer.render())
```

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by quagla** [comment]
Do the meshes contain only tetrahedra?

Cc @mohammad200h as he knows more about the GMSH format.

**Reply by Nick20500** [reply]
The statistics, which can be viewed with gmsh, confirms that there only should be tetrahedra.
![image](https://github.com/user-attachments/assets/1d487549-700b-430e-b654-3d28ff6edaf2)

**Comment by quagla** [comment]
Have you tried to set https://mujoco.readthedocs.io/en/latest/XMLreference.html#body-flexcomp-dim in the XML?

**Comment by mohammad200h** [comment, marked as answer]
You need to set dim=3 in xml. 
Here is a simplified model that loads.

[mujoco_model_solution.zip](https://github.com/user-attachments/files/17281915/mujoco_model_solution.zip)

if you look at the content of part-0.41.msh: 
```
$Nodes
1 1663 1 1663
3 0 0 1663
```
you can see 3 0 0 1663. 
3 indicates that this is a volumetric mesh. Therefore you need to set the dim=3 in xml.

**Reply by Nick20500** [reply]
This worked. 
![image](https://github.com/user-attachments/assets/1f491545-b061-45c2-8ada-9f408c4f11ef)
Thank you for your help and time.

Now I'm wondering why some xml files need the dimension explicitly stated and others not.
I have attached a folder with a banana and a bottle model.
The Bottle xml doesn't have the dimension set and can be loaded in the mujoco viewer without any errors.
Where as the Banana xml needs the dimension stated with 'dim="3"' for it to work without the "dim+1" error.
[mujoco_xml_dimension_examples.zip](https://github.com/user-attachments/files/17339046/mujoco_xml_dimension_examples.zip)

**Reply by mohammad200h** [reply]
This is due to a bug that will be fixed in the next release. 
What is happening is that dim defaults to 2 when it is not set.

in your bottle .msh file
```
$Elements
1 153 1 153
```
you have 153 elements. This is a volumetric mesh so each element is of size 4. 
```
$Elements
1 153 1 153
3 1 4 153
1 8 15 5 7 <- First element
2 3 1 9 17 
3 43 44 32 45 
```
so now you end up with  153*4 = 612
No look at this code from :
https://github.com/google-deepmind/mujoco/blob/476599f8b6175ef5a3040c47ebb3cd99a83c5365/src/user/user_flexcomp.cc#L225
```
  // check element size
  if (element.size()%(dim+1)) {
    return comperr(error, "Element size must be a multiple of dim+1", error_sz);
  }
```
```
element.size() =  153*4 = 612
dim =2
dim +1 = 3
element.size()%(dim+1) = 0 = Flase
```

Which means you don't get an error. basically you are getting lucky for the bottle but not for the banana.
The bug has been fixed and will be out in next release. Here are your current options:
1. set dim =3 in xml
2. wait for next release
3. build from source.
